agent/taskresource/cgroup/control/cgroupv2_controller_linux.go

```diff
@@ -0,0 +1,84 @@
+//go:build linux
+
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package control
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/cihub/seelog"
+	cgroupsv2 "github.com/containerd/cgroups/v2"
+)
+
+const (
+	defaultCgroupv2Path = "/sys/fs/cgroup"
+	parentCgroupSlice   = "/"
+	// This PID is only used when creating a slice for an existing systemd process.
+	// When creating a "general slice" that will be used as a parent slice for docker
+	// containers, then we use a dummy PID of -1.
+	// see https://github.com/containerd/cgroups/blob/1df78138f1e1e6ee593db155c6b369466f577651/v2/manager.go#L732-L735
+	generalSlicePID int = -1
+)
+
+// controlv2 is used to implement the cgroup Control interface
+type controlv2 struct{}
+
+// Create creates a new cgroup based off the spec post validation
+func (c *controlv2) Create(cgroupSpec *Spec) error {
+	// Validate incoming spec
+	err := validateCgroupSpec(cgroupSpec)
+	if err != nil {
+		return fmt.Errorf("cgroup create: failed to validate spec: %w", err)
+	}
+
+	cgroupPath := cgroupSpec.Root
+	seelog.Infof("Creating cgroup cgroupv2root=%s parentSlice=%s cgroupPath=%s", defaultCgroupv2Path, parentCgroupSlice, cgroupPath)
+	_, err = cgroupsv2.NewSystemd(parentCgroupSlice, cgroupPath, generalSlicePID, cgroupsv2.ToResources(cgroupSpec.Specs))
+	if err != nil {
+		return fmt.Errorf("cgroup create: unable to create v2 manager cgroupPath=%s err=%s", cgroupPath, err)
+	}
+
+	return nil
+}
+
+// Remove is used to delete the cgroup
+func (c *controlv2) Remove(cgroupPath string) error {
+	seelog.Infof("Removing cgroup cgroupv2root=%s parentSlice=%s cgroupPath=%s", defaultCgroupv2Path, parentCgroupSlice, cgroupPath)
+
+	m, err := cgroupsv2.LoadSystemd(parentCgroupSlice, cgroupPath)
+	if err != nil {
+		return err
+	}
+	return m.Delete()
+}
+
+// Exists is used to verify the existence of a cgroup
+func (c *controlv2) Exists(cgroupPath string) bool {
+	fullCgroupPath := filepath.Join(defaultCgroupv2Path, parentCgroupSlice, cgroupPath)
+	seelog.Infof("Checking existence of cgroup cgroupv2root=%s parentSlice=%s cgroupPath=%s fullPath=%s", defaultCgroupv2Path, parentCgroupSlice, cgroupPath, fullCgroupPath)
+
+	if _, err := os.Stat(fullCgroupPath); os.IsNotExist(err) {
+		return false
+	}
+	return true
+}
+
+// Init is used to setup the cgroup root for ecs
+func (c *controlv2) Init() error {
+	// In cgroups v2, there is no root "ecs" cgroup, so this function is a no-op.
```

Comment:
Parent cgroup aside, I think we will want to verify that the cpu and memory controllers are available under v2 (see discussion in this [section](https://quip-amazon.com/cJfYAHQJ5ZBM/Cgroupv2-Considerations-for-ECS-Agent#temp:C:OQQ24f5661d2d6beb0a39897f234)). If these two controllers are not available, docker won't be able to create container cgroups either. We currently don't have container resource limit as capability, so we won't be able to filter out the incapable instances during task placement. I personally think we should have that capability, but before that, we may want to terminate Agent since we can't respect task or container resource limits from task def in that case.

Reply:
Good point, I've added a check for the available cgroup controllers in the root /sys/fs/cgroup dir in the Init function now